File: Algorithm/Python/search_in_rotated_sorted_array_II.py
class Solution:
    def search(self,A,target):
        if not A:
            return False
        
        length=len(A)    
    
        l=0
        r=length-1
        while l<=r:
            mid=(l+r)/2
            print l,r
            if A[mid]==target:
                return True
            if A[mid]<A[r]: # the right is in order
                if A[mid]<target<=A[r]:
                    l=mid+1
                
                else:
                    r=mid-1         
            elif A[mid]>A[r]:  # the left is in order
                if A[l]<=target<A[mid]:
                    r=mid-1
                else:
                    l=mid+1
            else:
                r-=1
                
        return False
if __name__=='__main__':
    sol=Solution()
    import random
    A=[random.randint(0,45) for x in range(20)]
    A.sort()
    from collections import deque
    d=deque(A)
    d.rotate(44)
    A=list(d)
    A=[1,1,3,1]
    print A
       
    print sol.search(A,3)

File: Algorithm/CPP/search_in_rotated_sorted_array_ii.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <algorithm>
using namespace std;
class Solution {
public:
public:
	bool search(vector<int>& nums, int target)
	{
		int left = 0, right = nums.size()-1;
		while (left <= right)
		{
			int mid = (right - left) / 2 + left;
			if (nums[mid] == target) return true;
			if (nums[mid] < nums[right])
			{
				if (nums[mid] < target && target <= nums[right]) left = mid + 1;
				else right = mid - 1;

			}
			else if (nums[mid] > nums[right])
			{
				if (nums[left] <= target && target < nums[mid]) right = mid - 1;
				else left = mid + 1;
			}
			else
			{
				right -= 1;

			}
		}
		return false;
	}
};
int  main()
{
	Solution sol;
	vector<int> nums = { 4, 5, 6, 7, 8, 1, 2, 3,3 };
	int target =  8;
	//nums = { 5, 1, 3 };
	//target = 3;
	nums = { 3, 1 };
	target = 4;
	cout << sol.search(nums,target) << endl;
	cin.get();
	return 0;
}
